README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SILViA: Sample Incongruents Labeling and Visualization Approach

<!-- badges: start -->
<!-- badges: end -->

The goal of SILViA is to provide a method to identify outlying points
recorded by a [Valeport](https://www.valeport.co.uk) CTD device along a
depth profile, which are not congruent with the range of measurements
taken in the surrounding water column. These are labeled as
“incongruent” points.  
For this, the package includes a function to import and format .vp2
files, as well as functions to visualize depth profiles and the
incongruent points they contain.

## Installation

You can install the development version of SILViA from
[GitHub](https://github.com/) with:

``` r
install.packages("devtools")
devtools::install_github("KemenesT/SILViA")
library(SILViA)
```

## Usage

### Reading .vp2 files

The `read_vp2()` function is designed to read .vp2 files from Valeport
SWIFT CTD devices. The content of these files differs depending on
whether the device measures turbidity data or chlorophyll (through
fluorescence). Thus the function takes as input a directory to read
files from, the content that can be expected in the files, and the ID of
the device. This last input is recorded in the file name produced by a
SWIFT device.  
Files from different devices may thus be stored within the same
directory without preventing the function from reading the correct data.

``` r
# This function uses the package data `ctdat` to create example .vp2 files in 
# the temporary directory tempdir().
setup_example()

# The temporary files with example IDs "12345" for a chlorophyll-measuring CTD, 
# and "54321" for a turbidity-measuring CTD can be seen in the temporary 
# directory.
list.files(tempdir(), pattern = ".vp2") 
#> [1] "VL_12345_00000000000155b4609e492f.vp2"
#> [2] "VL_12345_00000000000255b43991379d.vp2"
#> [3] "VL_12345_00000000000355b427925258.vp2"
#> [4] "VL_12345_00000000000455b4774a1914.vp2"
#> [5] "VL_12345_00000000000555b4408627b0.vp2"
#> [6] "VL_54321_00000000000655b4de52b9f.vp2"

# Loading in the files from the CTD with ID "12345" can be done as:
chlorophyll_CTD_data <- read_vp2(directory = tempdir(), 
                                 type = "Chlorophyll", ID = "12345")
```

All the five different files are stacked in a single data frame as
follows:

|     | date       | time         | depth | pressure |       sv |   temp |    sal |     dens |   cond |  chla | filename                              |     lat |       lon |
|:----|:-----------|:-------------|------:|---------:|---------:|-------:|-------:|---------:|-------:|------:|:--------------------------------------|--------:|----------:|
| 2   | 2022/02/16 | 13:46:45.000 | 0.209 |    0.210 | 1540.338 | 26.882 | 36.408 | 1023.818 | 57.007 | 2.087 | VL_12345_00000000000155b4609e492f.vp2 | 30.5637 | -44.46387 |
| 3   | 2022/02/16 | 13:46:45.469 | 0.301 |    0.302 | 1540.348 | 26.810 | 36.571 | 1023.964 | 57.154 | 6.410 | VL_12345_00000000000155b4609e492f.vp2 | 30.5637 | -44.46387 |
| 4   | 2022/02/16 | 13:46:48.406 | 0.398 |    0.400 | 1540.321 | 26.689 | 36.804 | 1024.179 | 57.344 | 0.440 | VL_12345_00000000000155b4609e492f.vp2 | 30.5637 | -44.46387 |
| 5   | 2022/02/16 | 13:46:50.313 | 0.500 |    0.503 | 1539.953 | 26.679 | 36.477 | 1023.936 | 56.880 | 0.151 | VL_12345_00000000000155b4609e492f.vp2 | 30.5637 | -44.46387 |
| 6   | 2022/02/16 | 13:46:56.094 | 0.597 |    0.600 | 1539.909 | 26.670 | 36.453 | 1023.921 | 56.837 | 0.145 | VL_12345_00000000000155b4609e492f.vp2 | 30.5637 | -44.46387 |
| 7   | 2022/02/16 | 13:46:56.906 | 0.702 |    0.706 | 1539.951 | 26.680 | 36.469 | 1023.930 | 56.871 | 0.150 | VL_12345_00000000000155b4609e492f.vp2 | 30.5637 | -44.46387 |

``` r
# Note that only files from the desired CTD were imported:
unique(chlorophyll_CTD_data$filename)
#> [1] "VL_12345_00000000000155b4609e492f.vp2"
#> [2] "VL_12345_00000000000255b43991379d.vp2"
#> [3] "VL_12345_00000000000355b427925258.vp2"
#> [4] "VL_12345_00000000000455b4774a1914.vp2"
#> [5] "VL_12345_00000000000555b4408627b0.vp2"

# And that the data corresponds to chlorophyll measurements ("chla"):
colnames(chlorophyll_CTD_data)
#>  [1] "date"     "time"     "depth"    "pressure" "sv"       "temp"    
#>  [7] "sal"      "dens"     "cond"     "chla"     "filename" "lat"     
#> [13] "lon"

# To clear the temporary directory after using `setup_example()` use:
unlink(paste0(tempdir(), "\\", list.files(tempdir(), pattern = ".vp2")))
```

### Identifying Incongruent Points

The function `label_incongruents()` takes data as produced by
`read_vp2()` and adds two additional columns for every variable measured
by the CTD. Therefore, it also requires as input and expectation of the
kind of data that it receives. Here we specify:

- df1 = The chlorophyll data imported in the example above.  
- W = a depth window of 1m in which to consider the homogeneity of
  measurements.  
- alpha = a significance level of 0.1% for determining when an unlikely
  value is labeled as “incongruent”.  
- type = “Chlorophyll” so that the function knows to expect input data
  containing chlorophyll values.

``` r
labeled_data <- label_incongruents(df1 = chlorophyll_CTD_data, W = 1, 
                                   alpha = 0.001, type = "Chlorophyll")
```

The resulting data-frame contains new columns indicating for each
variable and each point whether it is labeled as incongruent
(‘Yes’/‘No’), and the p-value on which this decision is based. For
example:

|     | incongruent_temp |   pV_temp | incongruent_sal |    pV_sal |
|:----|:-----------------|----------:|:----------------|----------:|
| 2   | No               | 0.0150036 | No              | 0.8184680 |
| 3   | No               | 0.1425257 | No              | 0.3594200 |
| 4   | No               | 0.6728842 | Yes             | 0.0004519 |
| 5   | No               | 0.6786882 | No              | 0.6364872 |
| 6   | No               | 0.7086800 | No              | 0.7004968 |
| 7   | No               | 0.6416948 | No              | 0.6463254 |

### Plotting Profiles

Finally you can plot these results to visualize the labeled
incongruents.

The function `plot_profiles()` creates a “profiles.pdf” file in the
desired directory where each page contains plots for 6 variables
recorded by a CTD in a single cast: sound velocity, temperature,
salinity, density, conductivity, and either chlorophyll-a or turbidity.

Additionally, it outputs the data-frame with labeled incongruents used
to create these plots. This means that you can use `plot_profiles()`
directly if you want the labeled data frame and plots.

``` r
# After running this line check your working directory for the "profiles.pdf" file!
labeled_data <- plot_profiles(data = chlorophyll_CTD_data, width = 1, 
                              alpha = 0.001, type = "Chlorophyll", min.depth = 1.5)
#> A pdf with profiles was created in directory: C:/Users/thoma/Documents/SILViA

# To clear the pdf file created use the following line:
unlink(paste0(getwd(), "/profiles.pdf"))
```

The plots produced in the pdf look like this:

<img src="man/figures/README-unnamed-chunk-10-1.png" width="100%" />

Here the blue line is a moving average and red data points are those
labeled to be incongruent.


NEWS.md
# SILViA 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(SILViA)

test_check("SILViA")


R/incongruence_test.R
#' Incongruence Test
#'
#' Tests each point in the input vector for incongruence with the other elements
#' within the specified depth window.
#'
#' @param point_time Time measurement of the data point to be tested for
#' incongruence.
#'
#' @param column Name of the variable being labeled for incongruents in the
#' input data-frame "df1".
#'
#' @param subdf Subset data.frame inherited from [SILViA::run_inc_test()].
#'
#' @param W Halved window-length; inherited from [SILViA::label_incongruents()].
#'
#' @param alpha Level of significance used to test the incongruence of
#' data-points. Larger alpha-values make the function more sensitive in the
#' identification of incongruents.
#'
#' @param casts Vector of all cast names in the data-frame being processed;
#' inherited from [SILViA::run_inc_test()].
#'
#' @param cast Name of the active cast; inherited from
#' [SILViA::run_inc_test()].
#'
#' @importFrom stats sd
#' @importFrom stats pt
#'
#' @return Data frame containing the p-value and incongruence labeling for each
#' data point in the active cast.
#'
incongruence_test <- function(point_time, column, subdf, W,
                              alpha, casts, cast) {
  u <- which(subdf$time == point_time)
  active_depth <- subdf$depth[u]
  depth_range <- vector()
  depth_range[1] <- active_depth - W
  depth_range[2] <- active_depth + W
  window1 <- which(subdf$depth > depth_range[1] & subdf$depth < depth_range[2])

  if (length(window1) < 3) {
    message(paste0(
      "Window width is too narrow for point ", u, " in file ",
      casts[cast]
    ))
  }

  window1 <- window1[-which(window1 == u)]

  S <- sd(subdf[window1, column])
  Mean <- mean(subdf[window1, column])
  Tv <- (subdf[u, column] - Mean) / S
  pV <- pt(q = Tv, df = length(window1), lower.tail = FALSE)

  if (is.na(pV < alpha)) {
    incongruence <- "Error"
    pV <- "Error"
  } else if (pV < alpha) {
    incongruence <- "Yes"
  } else {
    incongruence <- "No"
  }

  output <- data.frame(incongruence = incongruence, pV = pV)
  return(output)
}


#' Run Incongruence Test
#'
#' Applies [SILViA::incongruence_test()] to each cast in a specified column from
#' the input data frame. This is done for the specified number of iterations.
#'
#' @param column Name of the variable being labeled for incongruents in the
#' input data-frame "df1".
#'
#' @param iterations Number of iterations for applying the incongruence labeling
#' test. Points labeled "incongruent" are ignored in subsequent iterations.
#'
#' @param df1 Data frame containing data from .vp2 files as obtained from
#' [SILViA::read_vp2()].
#'
#' @param W Halved window-length; inherited from [SILViA::label_incongruents()].
#'
#' @param alpha Level of significance used to test the incongruence of
#' data-points. Larger alpha-values make the function more sensitive in the
#' identification of incongruents.
#'
#' @return Data frame containing the original input data, followed by columns
#' indicating the incongruent labeling and p-value outcomes of each data point
#' for each variable.
#'
run_inc_test <- function(column, iterations, df1, W, alpha) {
  filename <- NULL

  for (iterate in 1:iterations) {
    subdf_all <- subset(df1, df1[, paste0("incongruent_", column)] == "No")

    active_df <- which(df1[, paste0("incongruent_", column)] == "No")

    casts <- unique(subdf_all$filename)

    for (cast in seq_along(casts)) {
      subdf <- subset(subdf_all, filename == casts[cast])
      active <- which(subdf_all$filename == casts[cast])

      output <- do.call(rbind.data.frame, lapply(
        X = subdf$time,
        FUN = incongruence_test,
        column, subdf, W, alpha,
        casts, cast
      ))

      df1[active_df[active], paste0(
        "incongruent_",
        column
      )] <- output$incongruence
      df1[active_df[active], paste0("pV_", column)] <- output$pV

      output <- df1[, c(paste0("incongruent_", column), paste0("pV_", column))]
    }
  }
  return(output)
}


tests/testthat/test-read_vp2.R
test_that("read_vp2 reads files correctly", {
  on.exit(unlink(paste0(
    tempdir(), "\\",
    list.files(tempdir(), pattern = ".vp2")
  )))
  setup_example()
  output <- read_vp2(tempdir(), type = "Chlorophyll", ID = 12345)
  expect_s3_class(output, class = "data.frame")
  expect_length(output, 13)
  expect_gt(nrow(output), 1615)
  expect_identical(
    colnames(output),
    c(
      "date", "time", "depth", "pressure", "sv", "temp", "sal",
      "dens", "cond", "chla", "filename", "lat", "lon"
    )
  )
  expect_type(output$lat, "double")
  expect_type(output$lon, "double")
  expect_false(any(is.na(output)))

  output <- read_vp2(tempdir(), type = "Turbidity", ID = 54321)
  expect_s3_class(output, class = "data.frame")
  expect_length(output, 15)
  expect_gt(nrow(output), 7)
  expect_identical(
    colnames(output),
    c(
      "date", "time", "depth", "pressure", "sv", "temp", "sal",
      "dens", "cond", "neph", "obs", "turb", "filename", "lat",
      "lon"
    )
  )
  expect_type(output$lat, "double")
  expect_type(output$lon, "double")
  expect_false(any(is.na(output)))

  expect_error(read_vp2(tempdir(), type = "other", ID = 54321),
    regexp = "'type' must be one of 'Chlorophyll' or 'Turbidity'"
  )
  expect_error(read_vp2(tempdir(), type = "Chlorophyll", ID = 54321),
    regexp = "Data type 'Chlorophyll' chosen for a device measuring Turbidity."
  )
  expect_error(read_vp2(tempdir(), type = "Turbidity", ID = 12345),
    regexp = "Data type 'Turbidity' chosen for a device measuring Chlorophyll."
  )
})

test_that("setup_example stores the example files in the temporary directory", {
  expect_type(ctdat, "list")
  on.exit(unlink(paste0(
    tempdir(), "\\",
    list.files(tempdir(), pattern = ".vp2")
  )))
  setup_example()
  expect_gt(length(sort(dir(path = tempdir(), pattern = ".vp2"))), 3)
  expect_identical(
    ctdat[[1]],
    readLines(paste0(
      tempdir(), "/",
      sort(dir(
        path = tempdir(),
        pattern = ".vp2"
      ))[1]
    ))
  )
})


R/read_vp2.R
#' Read .vp2 Files
#'
#' Read all .vp2 files in a directory given the CTD file ID and measurement
#' type.
#'
#' @param directory A character string of the directory from which to retrieve
#' .vp2 files.
#'
#' @param type A character string. Either "Chlorophyll" or "Turbidity" for the
#' appropriate CTD measurement type.
#'
#' @param ID Numeric. The ID of the CTD present in each file as:
#' "VL_ID_filenumber.vp2"
#'
#' @importFrom utils read.table
#'
#' @return A data frame containing the data for all .vp2 files in stacked
#' format.
#'
#' @export
#'
#' @examples
#' ## Load in example data and set it up in the temporary directory:
#' setup_example()
#'
#' ## Read the example casts:
#' casts <- read_vp2(directory = tempdir(), type = "Chlorophyll", ID = 12345)
#'
#' ## To clear the temporary directory after using 'setup_example()':
#' unlink(paste0(tempdir(), "/", list.files(tempdir(), pattern = ".vp2")))
#'
read_vp2 <- function(directory, type = c("Chlorophyll", "Turbidity"), ID) {
  files <- sort(dir(path = directory, pattern = ".vp2"))

  if (type == "Chlorophyll") {
    nms <- c(
      "date", "time", "depth", "pressure", "sv", "temp", "sal", "dens",
      "cond", "chla"
    )
  } else if (type == "Turbidity") {
    nms <- c(
      "date", "time", "depth", "pressure", "sv", "temp", "sal", "dens",
      "cond", "neph", "obs", "turb"
    )
  } else {
    stop("'type' must be one of 'Chlorophyll' or 'Turbidity'",
         call. = FALSE)
  }

  M <- as.data.frame(matrix(ncol = length(nms) + 3))

  names(M) <- c(nms, "filename", "lat", "lon")

  for (i in seq_along(files)) {
    fn <- files[i]
    if (substr(fn,
               nchar(fn) - 2, nchar(fn)) == "vp2" & substr(fn, 4, 8) == ID) {
      lns <- readLines(paste0(directory, "/", fn))
      if (length(lns) > 58) {
        if (suppressWarnings(
            is.na(as.numeric(strsplit(lns[27], "=")[[1]][2])))
            ) {
          lat <- as.numeric(strsplit(gsub(",", ".", lns[27]), "=")[[1]][2])
          lon <- as.numeric(strsplit(gsub(",", ".", lns[28]), "=")[[1]][2])
        } else {
          lat <- as.numeric(strsplit(lns[27], "=")[[1]][2])
          lon <- as.numeric(strsplit(lns[28], "=")[[1]][2])
        }

        if (lat != 999 | lon != 999) {
          if (type == "Chlorophyll") {
            if ("try-error" %in%
                class(try(read.table(paste0(directory, "/", fn), skip = 56)))) {
        stop("Data type 'Chlorophyll' chosen for a device measuring Turbidity.",
           call. = FALSE)
            }
            d <- read.table(paste0(directory, "/", fn), skip = 56)
          } else if (type == "Turbidity") {
            d <- read.table(paste0(directory, "/", fn), skip = 58)
            if (length(names(d)) != length(nms)) {
        stop("Data type 'Turbidity' chosen for a device measuring Chlorophyll.",
                   call. = FALSE)
            }
          }
          names(d) <- nms
          d$filename <- fn
          d$lat <- lat
          d$lon <- lon
          M <- rbind(M, d)
        }
      }
    }
  }

  M <- M[-1, ] # Removes empty row

  return(M)
}


#' Setup Example CTD Files
#'
#' Sets up the files from ctdat.RData in a temporary directory to read with
#' [SILViA::read_vp2()].
#'
#' @export
#' @examples
#' setup_example() # Saves example CTD files in a temporary directory
#'
#' ## These can be read as:
#' casts <- read_vp2(directory = tempdir(), type = "Chlorophyll", ID = 12345)
#'
#' ## To clear the temporary directory after using 'setup_example()':
#' unlink(paste0(tempdir(), "/", list.files(tempdir(), pattern = ".vp2")))
#'
setup_example <- function() {
  file1 <- tempfile("VL_12345_000000000001",
    fileext = ".vp2"
  )
  file2 <- tempfile("VL_12345_000000000002",
    fileext = ".vp2"
  )
  file3 <- tempfile("VL_12345_000000000003",
    fileext = ".vp2"
  )
  file4 <- tempfile("VL_12345_000000000004",
    fileext = ".vp2"
  )
  file5 <- tempfile("VL_12345_000000000005",
    fileext = ".vp2"
  )
  file6 <- tempfile("VL_54321_000000000006",
    fileext = ".vp2"
  )
  writeLines(ctdat[[1]], con = file1)
  writeLines(ctdat[[2]], con = file2)
  writeLines(ctdat[[3]], con = file3)
  writeLines(ctdat[[4]], con = file4)
  writeLines(ctdat[[5]], con = file5)
  writeLines(ctdat[[6]], con = file6)
}


R/moving_fns.R
#' Moving Function
#'
#' Applies a specified function to subsets of a vector sequentially to compute a
#' 'moving window' along the the vector.
#' The window is defined by each element as either the subsequent (right),
#' preceding (left), or neighboring (center) elements.
#'
#' @param x Vector with numeric data.
#' @param w Window length.
#' @param fun Function to apply.
#' @param side Side to take, (c)entre, (l)eft or (r)ight.
#' @param ... Parameters passed on to 'fun'.
#'
#' @return Vector containing the function outputs for each window along the
#' vector x.
#'
#' @export
#'
#' @examples
#' my_data <- rnorm(50, mean = 30, sd = 5)
#' moving_averages <- moving_fn(my_data, w = 5, fun = mean, side = "center")
#'
moving_fn <- function(x, w, fun, side, ...) {
  y <- vector(mode = class(x), length = length(x))
  for (i in seq_len(length(x))) {
    if (side %in% c("c", "centre", "center")) {
      ind <- c((i - floor(w / 2)):(i + floor(w / 2)))
    } else if (side %in% c("l", "left")) {
      ind <- c((i - floor(w) + 1):i)
    } else if (side %in% c("r", "right")) {
      ind <- c(i:(i + floor(w) - 1))
    } else {
      stop("'side' must be one of 'centre', 'left', 'right'", call. = FALSE)
    }
    ind <- ind[ind %in% seq_len(length(x))]
    y[i] <- fun(x[ind], ...)
  }
  return(y)
}


R/plot_profiles.R
#' Plot CTD Profiles
#'
#' Plots CTD profiles for all variables measured by the CTD are included. Data
#' identified as incongruent are shown as red points, otherwise they are black.
#' The blue line is computed as the moving average with a window of 5 data
#' points constructed with [SILViA::moving_fn()], ignoring data identified as
#' incongruent.
#'
#' @param data Data frame containing data from .vp2 files as obtained from
#' [SILViA::read_vp2()].
#'
#' @param width Window width against which each data-point is compared along the
#' CTD depth profile. Units should match those expressed in the "depth" column
#' of the input data.
#'
#' @param alpha Level of significance used to test the incongruence of
#' data-points. Larger alpha-values make the function more sensitive in the
#' identification of incongruents.
#'
#' @param iterations Number of iterations for applying the incongruence labeling
#' test. Points labeled "incongruent" are ignored in subsequent iterations.
#'
#' @param type A character string. Either "Chlorophyll" or "Turbidity" for the
#' appropriate CTD measurement type.
#'
#' @param min.depth Minimum depth above which points are excluded from the
#' profiles. Units should match those expressed in the "depth" column of the
#' input data.
#'
#' @param directory The directory in which to save the "profiles.pdf" file. The
#' working directory is used by default.
#'
#' @param plots Vector of character strings indicating which variables to use as
#' criteria for selecting profiles to plot in the final pdf. All the profiles
#' from a cast will be plotted if an incongruent data point was labeled for the
#' selected variables. The default "any" only excludes plots from casts that
#' had no incongruent data points for all variables. The option "all" plots
#' all casts regardless of whether they had incongruent data points.
#'
#' @return Data frame containing the input data-frame with columns labeling the
#' incongruence of each point for each variable; the "warning" column indicating
#' whether a certain measurement depth has incongruent values for more than two
#' variables. A "profiles" pdf is also created in the working directory
#' including plotted profiles for which Chlorophyll or Turbidity points were
#' identified as incongruent. Red points indicate those labeled as incongruent.
#'
#' @importFrom grDevices dev.off
#' @importFrom grDevices pdf
#' @importFrom gridExtra grid.arrange
#' @importFrom grid textGrob
#' @importFrom grid gpar
#'
#' @export
#'
#' @examples
#' ## Load in example data and set it up in the temporary directory:
#' setup_example()
#'
#' ## Read the example casts:
#' casts <- read_vp2(directory = tempdir(), type = "Chlorophyll", ID = 12345)
#'
#' ## Select a few samples to plot quickly:
#' selected_files <- unique(casts$filename)[3:4]
#' casts <- casts[which(casts$filename %in% selected_files),]
#'
#' ## Label incongruents in the imported data and create a pdf with appropriate
#' ## plots:
#' output <- plot_profiles(
#'   data = casts, width = 0.6, alpha = 0.001,
#'   type = "Chlorophyll", min.depth = 1.5,
#'   directory = tempdir()
#' )
#'
#' ## To delete the example pdf you can run the following:
#' unlink(paste0(tempdir(), "/profiles.pdf"))
#'
#' ## To clear the temporary directory after using 'setup_example()':
#' unlink(paste0(tempdir(), "\\", list.files(tempdir(), pattern = ".vp2")))
#'
plot_profiles <- function(data, width, alpha, iterations = 1,
                          type = c("Chlorophyll", "Turbidity"), min.depth = 0,
                          directory = getwd(), plots = "any") {
  type <- match.arg(type)
  pV_sv <- pV_temp <- pV_sal <- pV_ <- pV_dens <- pV_cond <- pV_chla <-
    pV_neph <- pV_obs <- pV_turb <- NULL
  data <- data[data$depth > min.depth, ]

  if (type == "Chlorophyll") {
    data <- label_incongruents(
      df1 = data, W = width, alpha = alpha,
      iterations = iterations, type = type
    )

    if (plots == "all") {
      files <- unique(data$filename)
    } else if (plots == "any") {
      plots <- c("sv", "temp", "sal", "dens", "cond", "chla")
      plots <- paste0("incongruent_", plots)
      files <- unique(data[which(apply(data[plots] == "Yes", 1, any)),
                           ]$filename)
    } else {
      plots <- paste0("incongruent_", plots)
      files <- unique(data[which(apply(data[plots] == "Yes", 1, any)),
                           ]$filename)
    }

    pdf(paste0(directory, "/profiles.pdf"), width = 8.3, height = 11.7)

    for (f in seq_along(files)) {
      do1 <- data[data$filename == files[f], ]
      nms <- c(
        "date", "time", "depth", "pressure", "sv", "temp", "sal", "dens",
        "cond", "chla"
      )

      profiles <- lapply(c("sv", "temp", "sal", "dens", "cond", "chla"),
        draw_plot,
        data = do1
      )

      suppressWarnings(grid.arrange(profiles[[1]], profiles[[2]], profiles[[3]],
        profiles[[4]], profiles[[5]], profiles[[6]],
        left = textGrob("Depth (m)", rot = 90, gp = gpar(fontsize = 13)),
        top = textGrob(files[f], gp = gpar(fontsize = 15)),
        nrow = 2, ncol = 3
      ))
    }

    dev.off()
    message(paste0("A pdf with profiles was created in directory: ", directory))
    return(data)
  } else if (type == "Turbidity") {
    data <- label_incongruents(
      df1 = data, W = width, alpha = alpha,
      iterations = iterations, type = type
    )

    if (plots == "all") {
      files <- unique(data$filename)
    } else if (plots == "any") {
      plots <- c("sv", "temp", "sal", "dens", "cond", "turb")
      plots <- paste0("incongruent_", plots)
      files <- unique(data[which(apply(data[plots] == "Yes", 1, any)),
                           ]$filename)
    } else {
      plots <- paste0("incongruent_", plots)
      files <- unique(data[which(apply(data[plots] == "Yes", 1, any)),
                           ]$filename)
    }

    pdf(paste0(directory, "/profiles.pdf"), width = 8.3, height = 11.7)

    for (f in seq_along(files)) {
      do1 <- data[data$filename == files[f], ]
      nms <- c(
        "date", "time", "depth", "pressure", "sv", "temp", "sal", "dens",
        "cond", "neph", "obs", "turb"
      )

      profiles <- lapply(c("sv", "temp", "sal", "dens", "cond", "turb"),
        draw_plot,
        data = do1
      )

      suppressWarnings(grid.arrange(profiles[[1]], profiles[[2]], profiles[[3]],
        profiles[[4]], profiles[[5]], profiles[[6]],
        left = textGrob("Depth (m)", rot = 90, gp = gpar(fontsize = 13)),
        top = textGrob(files[f], gp = gpar(fontsize = 15)),
        nrow = 2, ncol = 3
      ))
    }

    dev.off()
    message(paste0("A pdf with profiles was created in directory: ", directory))
    return(data)
  }
}


#' Plot Profile
#'
#' Plots the profile for a selected variable from the CTD data-set. The plot
#' consists of all data points, with those labeled as incongruent colored red.
#' The moving average for a window length of 5 points constructed with
#' [SILViA::moving_fn()] is shown as a blue line.
#'
#' @param var Name of the variable to be plotted. This should be one of: "sv",
#' "temp", "sal", "dens", "cond", "chla", "neph", "obs", or "turb".
#' @param data Data containing the variable to be plotted. This must be a
#' data-frame as returned by [SILViA::label_incongruents()] or
#' [SILViA::plot_profiles()].
#'
#' @return ggplot object of the selected variable.
#'
#' @importFrom dplyr %>%
#' @importFrom dplyr arrange
#' @importFrom ggplot2 ggplot
#' @importFrom ggplot2 aes
#' @importFrom ggplot2 geom_point
#' @importFrom ggplot2 geom_path
#' @importFrom ggplot2 labs
#' @importFrom ggplot2 theme_linedraw
#' @importFrom ggplot2 element_text
#' @importFrom ggplot2 element_blank
#' @importFrom ggplot2 theme
#' @importFrom ggplot2 ylab
#' @importFrom ggplot2 xlab
#'
#' @export
#'
#' @examples
#' ## Load in example data and set it up in the temporary directory:
#' setup_example()
#'
#' ## Read the example casts:
#' casts <- read_vp2(directory = tempdir(), type = "Chlorophyll", ID = 12345)
#'
#' ## Select the data for one file:
#' selected_filename <- unique(casts$filename)[5]
#' single_file <- casts[which(casts$filename == selected_filename), ]
#'
#' ## Label incongruents in the data:
#' output <- label_incongruents(
#'   df1 = single_file, W = 1, alpha = 0.0001,
#'   type = "Chlorophyll"
#' )
#'
#' ## Plot a single profile:
#' draw_plot(var = "chla", data = output)
#'
#' ## To clear the temporary directory after using 'setup_example()':
#' unlink(paste0(tempdir(), "\\", list.files(tempdir(), pattern = ".vp2")))
#'
draw_plot <- function(var, data) {
  y <- depth <- NULL

  if (identical(which(data[paste0("incongruent_",
                                  var)] == "Yes"), integer(0))) {
    d <- data
  } else {
    d <- data[-which(data[paste0("incongruent_", var)] == "Yes"), ]
  }

  sel <- as.numeric(order(d$time))
  sel2 <- as.numeric(order(data$time))
  d <- as.data.frame(d)
  data <- as.data.frame(data)
  moving_results <- data.frame(
    x = moving_fn(
      x = d[sel, var], w = 5, fun = mean, side = "centre",
      na.rm = FALSE
    ),
    y = -moving_fn(
      x = d[sel, "depth"], w = 5, fun = mean, side = "centre",
      na.rm = FALSE
    )
  ) %>% arrange(y)

  profile <- ggplot() +
    geom_point(
      aes(x = data[sel2, var], y = -data[sel2, "depth"]),
      color = "red", pch = 16, cex = .6
    ) +
    geom_point(
      aes(x = d[sel2, var], y = -d[sel2, "depth"]),
      color = "black", pch = 16, cex = .6
    ) +
    geom_path(aes(x = moving_results$x, y = moving_results$y), color = "blue") +
    labs(title = var) +
    theme_linedraw() +
    theme(
      plot.title = element_text(face = "bold", vjust = 0.01),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.text.x = element_text(angle = -25, vjust = 0.005, size = 7)
    ) +
    ylab(label = "") +
    xlab(label = "")

  return(profile)
}
